Fix Move jumpy-ness and occasional facing mismatch

reaperrr opened the pull request:
1) Fixes facing mismatch bug:
Due to `var ret` passing `mobile.Facing` to the next queued `MovePart` before `UpdateCenterLocation` would set the correct final facing of the current `MovePart`, this was preventing the correct 'final' facing from being set as early as it should, making actors only turn to the correct final facing at the very end of `MoveSecondHalf`.

Best testcase I can offer is the harvester in the RA shellmap: Watch it closely when it makes its final turn towards the first cell it harvests. If you compare bleed and this PR, you should be able to see the difference.

Fixes #16651.
Fixes #18732 (which admittedly was just a duplicate of the above).

2) Fixes ground actors moving too fast and 'jump' during MovePart transitions:
`MovePart` didn't account for usually being on the same tick as the previous `MovePart`s last tick if the latter overshot `MoveFractionTotal` (which is usually the case).

A ground actor with base speed 100 on terrain with 100% modifier will now move exactly as fast as an aircraft with the same speed.

Fixes #13844 (tested)
Fixes #16682.

reaperrr committed:
Simplify MovePart code

InnerActivity and UpdateCenterLocation made this
overly complex and hard to read & debug.

This also fixes a bug that would make an outdated
facing being passed during OnComplete (because
InnerActivity was cached before UpdateCenterLocation
could set the correct final facing).

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -485,7 +485,7 @@ void UpdateCenterLocation(Actor self, Mobile mobile)
 				if (moveFraction >= MoveFractionTotal)
 					mobile.Facing = ToFacing;
-				else
+				else if (FromFacing != ToFacing)
```
I suspect that this is putting a bandaid over some other bug, which I expect will be breaking something that we don't realize yet!

If this were just a rounding issue then I would expect that the following would behave the same as the current PR:
```csharp
if (moveFraction >= MoveFractionTotal)
	mobile.Facing = ToFacing;
else if (FromFacing == ToFacing)
	mobile.Facing = ToFacing;
else
	mobile.Facing = WAngle.Lerp(FromFacing, ToFacing, moveFraction, MoveFractionTotal);
```
However, it actually keeps the behaviour that is on bleed, which probably means that this part of the code is behaving correctly, but that `ToFacing` has an unexpected value &rarr; maybe related to the classic facing fudge?

reaperrr replied:
I think the problem was that `var ret` would make the `OnComplete` pass an outdated `mobile.Facing` to the next MovePart, since `var ret` is created before `UpdateCenterLocation` runs.

Updated, hopefully this is the correct fix now.

pchote commented:
That looks like it should work, but IMO it highlights the current awkward split of logic we have between `Tick`, `InnerTick`, and `UpdateCenterLocation`.

Would there be any downside to combining these together into something like:
```csharp
public override bool Tick(Actor self)
{
	var mobile = Move.mobile;
	moveFraction += mobile.MovementSpeedForCell(self, mobile.ToCell);

	if (moveFraction >= MoveFractionTotal)
	{
		moveFraction = MoveFractionTotal;
		mobile.SetCenterPosition(self, To);
		mobile.Facing = ToFacing;

		Queue(OnComplete(self, mobile, Move));
		return true;
	}

	WPos pos;
	if (EnableArc)
	{
		var angle = WAngle.Lerp(ArcFromAngle, ArcToAngle, moveFraction, MoveFractionTotal);
		var length = int2.Lerp(ArcFromLength, ArcToLength, moveFraction, MoveFractionTotal);
		var height = int2.Lerp(From.Z, To.Z, moveFraction, MoveFractionTotal);
		pos = ArcCenter + new WVec(0, length, height).Rotate(WRot.FromYaw(angle));
	}
	else
		pos = WPos.Lerp(From, To, moveFraction, MoveFractionTotal);

	if (self.Location.Layer == 0)
		pos -= new WVec(WDist.Zero, WDist.Zero, self.World.Map.DistanceAboveTerrain(pos));

	mobile.SetCenterPosition(self, pos);
	mobile.Facing = WAngle.Lerp(FromFacing, ToFacing, moveFraction, MoveFractionTotal);
	return false;
}
```
?

IMO this makes the code a lot easier to understand, any maybe easier to fix the bug where the MoveFraction overshoot is lost (just need to pass `moveFraction - moveFractionTotal` to `OnComplete`).

reaperrr committed:
Fix Move jumpy-ness on MovePart transitions

There were 2 issues at work here, both when progress
would overshoot Distance (which is the usual case,
rather than the exception):
- the overshot progress was passed on by MoveFirstHalf, however
  OnComplete would make the next MovePart start ticking the
  same tick on which the old MovePart reached Distance,
  but move by carryoverProgress +(!!!) terrain speed instead of moving
  by just the left-over carryoverProgress.
- MoveSecondHalf would not pass any overshot progress to the
  next MoveFirstHalf queued by parent Move, leading to
  the next MoveFirstHalf performing a full-speed move the same tick
  MoveSecondHalf finished its last move.

reaperrr commented:
Updated.

Also managed to fix the jumpy transitions between `MovePart`s. Hopefully without accidentally regressing anything I overlooked.

pchote commented:
This again looks like it should work but highlights awkwardness in the existing code. I was initially going to complain about carrying over `startingFraction` through the parent `Move`, but then I realized that my main issue was the incorrect implication from `starting`, and then i realized that these aren't even fractions at all!

Would it be too annoying if I ask to add a new commit either before or after the "Simplify MovePart code" commit that renames `moveFraction` &rarr; `progress` and `MoveFractionTotal` &rarr; `Distance` (or something different if you have an idea for better names)? `Move.startingFraction` could then be renamed to `Move.carryoverProgress` in the last commit, which IMO would make the rest of the commit a lot easier to understand.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -437,11 +442,17 @@ public MovePart(Move move, WPos from, WPos to, WAngle fromFacing, WAngle toFacin
 				var mobile = Move.mobile;
-				moveFraction += mobile.MovementSpeedForCell(self, mobile.ToCell);
+
+				// If previous MovePart overshot MoveFractionTotal and we carried over a
```
It took me a couple of goes to understand this. IMO a clearer way to present this would be to replace `needToMoveByStartingFraction` with a `firstTick` bool, and then change this to:
```csharp
// Having non-zero progress in the first tick means that this MovePart is following on from
// a previous MovePart that has just completed during the same tick. In this case, we want to
// apply the carried over progress but not evaluate a full new step until the next tick.
if (!firstTick || progress == 0)
	progress += mobile.MovementSpeedForCell(self, mobile.ToCell);

firstTick = false;
```

What do you think?

pchote commented:
Also, very nice detective work! The double-ticking versus ignored carryover explains why the aircraft versus mobile speed resutls were so inconsistent when people were previously testing this...

This is going to break the balancing in RA and TD until the community can test and tweak values, but this fix has been a long time coming so shouldn't be delayed any further.

Smittytron commented:
> The double-ticking versus ignored carryover explains why the aircraft versus mobile speed resutls were so inconsistent when people were previously testing this...

Would this also close #16682 then?

pchote commented:
Yes. I've edited this into the PR description to link the auto-closes.

reaperrr committed:
Rename fraction to distance/progress in MovePart

reaperrr commented:
Updated.

Inq8 commented:
I need to do further testing of this, but when I cherry-picked just this PR & #19219 onto release I had substantial FPS loss (15-25) on an 8 AI map replay with lots of units present.

Is this significantly more costly per unit?

reaperrr commented:
@Inq8 To be honest, I don't see how this should affect performance at all.

reaperrr commented:
> To be honest, I don't see how this should affect performance at all.

Which doesn't completely rule out that we're overlooking something, mind you.
But the only (indirect) remotely notable effect this _should_ have on performance is that without adjusted `Mobile.Speed`, ground units will simply need a few more ticks to reach their destination now, effectively spending more time on movement.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -436,57 +441,43 @@ public MovePart(Move move, WPos from, WPos to, WAngle fromFacing, WAngle toFacin
-					}
-					else
-						pos = WPos.Lerp(From, To, moveFraction, MoveFractionTotal);
+					progress = Distance;
```
This resets the carryover progress to 0, undoing the fix!

I can confirm that after removing this line unit movement does finally work as expected.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -562,12 +553,15 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
 			{
 				mobile.SetPosition(self, mobile.ToCell);
+
+				// Move might immediately queue a new MoveFirstHalf, so we need to pass a carryoverProgress there too
```
I find this comment a bit confusing - this is the only place that sets `carryoverProgress`, but the "there too" suggests there are more.

I also found it difficult to clearly explain this, but eventually came up with this:
```csharp
// Move might immediately queue a new MoveFirstHalf within the same tick if we haven't
// reached the end of the requested path. Make sure that any leftover movement progress is
// correctly carried over into this new activity to avoid a glitch in the apparent move speed.
```

reaperrr commented:
Updated.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d58fbe46e760328e39508a9ea8bde03e822177ca...d0d93428bbc57271e362b0bad400f1d0d4f2c7df
